src/gc.c

```diff
@@ -1300,11 +1283,19 @@ STATIC_INLINE jl_value_t *jl_gc_pool_alloc_inner(jl_ptls_t ptls, int pool_offset
         if (v != NULL) {
             // like the freelist case,
             // but only update the page metadata when it is full
-            jl_gc_pagemeta_t *pg = jl_assume(page_metadata((char*)v - 1));
+            jl_gc_pagemeta_t *pg = jl_assume(page_metadata_unsafe((char*)v - 1));
             assert(pg->osize == p->osize);
             pg->nfree = 0;
             pg->has_young = 1;
-            v = *(jl_taggedvalue_t**)cur_page;
+            pg = pop_page_metadata_back(&ptls->page_metadata_lazily_freed);
+            if (pg != NULL) {
+                v = gc_reset_page(ptls, p, pg);
```

Comment:
Shouldn't this page still have real objects on it? What happened to the partially allocated pages when you reset them?

Reply:
`pg` will be assigned to the metadata of a lazily freed page, which, by definition, shouldn't have any allocated objects.

Reply:
And why are you now ignoring page `cur_page`, instead of doing allocations to it

Reply:
We reach this code block if we have exhausted `p->freelist` and if the next attempt of bump allocating from `p->newpages` would go beyond the page boundary, which means we can no longer allocate on this page.

Reply:
This page is indeed empty now, but the next page (previously from `*cur_page`) still had space left

Reply:
This code seems to be in the wrong place, since it is duplicating the work of `gc_add_page`, but not in the branch with it

Reply:
> This code seems to be in the wrong place, since it is duplicating the work of gc_add_page, but not in the branch with it

Yeah, this logic should likely be moved into `gc_add_page`.

> This page is indeed empty now, but the next page (previously from *cur_page) still had space left

The page layout has been changed so that it stores a pointer to the page metadata in the first word of the page. The `next` pointer that previously stored a pointer to lazily freed pages no longer exists, and has been effectively replaced by `ptls->page_metadata_lazily_freed`.

Reply:
The next pointer was for the next newpage, not the next lazily freed page

Reply:
Seems like we were only assigning a non-NULL value for the `next` pointer when we were lazily freeing the page?

E.g.: https://github.com/JuliaLang/julia/blob/ad120f49bd606d91351e71fd5e5261daf9a981f9/src/gc.c#L1255 and https://github.com/JuliaLang/julia/blob/ad120f49bd606d91351e71fd5e5261daf9a981f9/src/gc.c#L1369

Reply:
But that doesn't make sense. There is no freelist on the newpage (just random garbage bits after the end of the bump-pointer section), so you can't just walk through that turning it into a freelist, as you may misinterpret undef data as allocated objects after they change pool size.

(Apparently though we do not preserve halfpages in the newlist after the first sweep, but rather move them eagerly to the slower freelist, so that is curious. I feel I was misinformed about that.)

Reply:
> There is no freelist on the newpage (just random garbage bits after the end of the bump-pointer section), so you can't just walk through that turning it into a freelist, as you may misinterpret undef data as allocated objects after they change pool size.

That's a great observation and could indeed lead us to preserve objects that are actually garbage. I can think of two solutions:

1. `memset` the page to 0 upon allocation.
2. Only walk the page up to the cursor defined by `p->newpages`, and don't reset `p->newpages` after sweeping. By only walking the page up to `p->newpages` we should guarantee to only observe valid objects.

Reply:
The second solution could introduce some fragmentation, though, as objects before `p->newpages` might actually have been freed...

Reply:
Option 2 is what the gc-debug implements and uses

Reply:
But since we are already destroying the bump-pointer allocator at the first quick-sweep, maybe it doesn't matter?

Reply:
https://github.com/JuliaLang/julia/pull/50357